agent/internal/task_config.go

```diff
@@ -52,10 +53,46 @@ type TaskConfig struct {
 	ModulePaths        map[string]string
 	CedarTestResultsID string
 	TaskGroup          *model.TaskGroup
+	CommandCleanups    []CommandCleanup
 
 	mu sync.RWMutex
 }
 
+type CommandCleanupLevel int
+
+const (
+	// TaskCleanupLevel runs the cleanup commands after the task is finished.
+	// The task is considered 'pre+main+post' or 'setup_task+main+teardown_task'
+	// and includes timeout commands if the task timed out.
+	TaskCleanupLevel = CommandCleanupLevel(iota)
+	// TaskGroupCleanupLevel runs the cleanup commands after the task group is finished.
+	// This is for 'setup_group+teardown_group'.
+	TaskGroupCleanupLevel
+)
+
+// CommandCleanups is a list of cleanup functions that are added dynamically
+// during task execution. These functions are called when the task is
+// finished and then cleared from the list. A 'task' in this context is what
```

Comment:
when the task is finished -> when the command has gone out of scope? Phrasing is a little ambiguous here.

Maybe we can split by "normal task" vs. "task-group-specific" command blocks? All the normal task blocks and task-group-equivalent blocks are cleaned up when the task ends, whereas setup group and teardown group are the only ones that clean up when the task group ends.